oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java

```diff
@@ -105,6 +109,9 @@ private OAuth2AuthorizationRequest resolve(HttpServletRequest request, String re
 		OAuth2AuthorizationRequest.Builder builder;
 		if (AuthorizationGrantType.AUTHORIZATION_CODE.equals(clientRegistration.getAuthorizationGrantType())) {
 			builder = OAuth2AuthorizationRequest.authorizationCode();
+			if (isEmpty(clientRegistration.getClientSecret())) {
```

Comment:
Is this the best way to determine if PKCE should be used? This seems implicit rather than explicit which can lead to problems later on (if there are other reasons the client secret can be empty). Additionally implicit options in security tends to lead to security vulnerabilities.

Reply:
I think this makes sense. If `AuthorizationGrantType.AUTHORIZATION_CODE` and `ClientAuthenticationMethod.NONE` (public client) than PKCE parameters can be applied.

Reply:
@jgrandja I like the `AuthorizationGrantType.AUTHORIZATION_CODE` and `ClientAuthenticationMethod.NONE` combination to determine PKCE usage. @rwinch I agree the absence of a client secret is not great. Are you ok with `AuthorizationGrantType.AUTHORIZATION_CODE` + `ClientAuthenticationMethod.NONE`?

Reply:
Hi @jgrandja and @rwinch, I received an email from someone watching this PR that raises a good point. Attaching below:

> Hi Stephen,
> 
> Recently I bumped into your pullrequest for PKCE.
> 
> For our project we even have the requirement of using PKCE for normal web applications as an extra hardening.
> 
> Okta has put a demo of the PKCE in their playground:
> 
> https://oauth.com/playground/authorization-code-with-pkce.html
> 
> You will see that in the final step, to get the token, they pass both the client_secret (client authentication client_secret_basic in this case) and code_verifier.
> 
> 
> In the pull request I saw some discussion on how to determine PKCE:
> 
> https://github.com/spring-projects/spring-security/pull/6485#discussion_r253276486
> 
> In our case or the above mentioned example on Okta this would not be fine. I believe we need a special parameter where we can force PKCE.
> 
> What do you think?

I think he makes a good point but it would mean some kind of further flag(?) on ClientRegistration like "withPkce" (true/false) that would only apply for authorization_code grant type. This way PKCE could be used with or without a client secret.

Reply:
@sdoxsee I'm not sure I'm following the thread in the email you received.

PKCE is meant to be used for Public Clients leveraging the Authorization Code grant. Public Clients are not assigned full credentials (no client secret). So I guess I'm confused with what the user is saying.

Looking at the Okta developer guide for [Authorization Code Flow with PKCE](https://developer.okta.com/authentication-guide/implementing-authentication/auth-code-pkce), you will see that the client authentication is not required (so no `client_secret`).

> Important: Unlike the regular Authorization Code Flow, this call does not require the Authorization header with the client ID and secret. This is why this version of the Authorization Code flow is appropriate for native apps.

And for the standard [Authorization Code Flow](https://developer.okta.com/authentication-guide/implementing-authentication/auth-code), you will see that the client authentication is required as this flow is targeted for confidential clients.

> Important: The call to the /token endpoint requires authentication. In this case, it is a Basic Auth digest of the client ID and secret. You can find the client ID and secret in your application’s General tab. This requirement is why this call is only appropriate for applications that can guarantee the secrecy of the client secret. For more on Basic Auth, please see Token Authentication Methods.

Reply:
@jgrandja. I agree that it is not typical to compose authorization_code grant (with `client_secret`) AND PKCE but, despite the PKCE spec being "Proof Key for Code Exchange by OAuth **Public Clients**", the spec doesn't seem to prohibit the use of `client_secret`. For instance, the [token endpoint request for PKCE](https://tools.ietf.org/html/rfc7636#section-4.5) just builds on the generic [OAuth2 authorization_code token request](https://tools.ietf.org/html/rfc6749#section-4.1.3). 

You're right, for public clients, client authentication is not _required_ but it's not prohibited. In any case, this is not an important customization for my use case. I've contacted the person who emailed to chime in if it is important for them.

Reply:
@sdoxsee In reference to...

> For instance, the [token endpoint request for PKCE](https://tools.ietf.org/html/rfc7636#section-4.5) just builds on the generic [OAuth2 authorization_code token request](https://tools.ietf.org/html/rfc6749#section-4.1.3).

If you dig deeper into section, 4.1.3.  Access Token Request

> **If the client type is confidential** or the client was issued client
   credentials (or assigned other authentication requirements), the
   client MUST authenticate with the authorization server as described
   in Section 3.2.1.

So for a public client (PKCE), which is never issued a `client_secret`, since it cannot securely store it (e.g. a native app) it could not authenticate anyway since it does not have a `client_secret` assigned during registration. See [Client Types](https://tools.ietf.org/html/rfc6749#section-2.1)

Reply:
Thanks @jgrandja. Yes, I drilled down to that definition as well. What this other person is saying is that they have a **confidential** client (likely a traditional oauth2 authorization code flow web app?) but want to add PKCE also--as extra security. In such a case, you WOULD have a PKCE client that HAS a `client_secret`

Why? I'm not sure but I can only imagine that it's to protect against the scenario where the "code" comes back via the front channel and they want to prevent some malicious agent, that intercepts the "code" and somehow has not only the client_id but also the client_secret, from getting tokens. If their `client_secret` is compromised, I believe they probably have bigger problems. However, the person who brought this up is saying that they have this requirement to add PKCE to a confidential client and I can't see why they should not be able to do it according to the spec.

Reply:
Thanks for the explanation @sdoxsee. Now I understand the use case. However, this seems more of an edge case to me. I'd prefer that we leave this out of this PR so we can get this merged. Can you please add a new ticket so we can gather more information.

> the person who brought this up is saying that they have this requirement to add PKCE to a confidential client

Can you also please have the user respond to the new ticket with more detail and specifically which Authorization Server supports PKCE for a confidential client (Okta?). A link to server docs would be even better.

Reply:
Hi,
The current best practices for OAuth security specifies that PKCE should also be used for webapps:

> Note: although PKCE so far was recommended as a mechanism to protect
   native apps, this advice applies to all kinds of OAuth clients,
   **_including web applications_**.

see: https://tools.ietf.org/html/draft-ietf-oauth-security-topics-11#section-2.1.1

Now that you are implementing PKCE you better foresee that PKCE can also be used for web applications.

Reply:
@vanwobe Thanks for the reference. Just below the reference it continues with...

> Authorization servers MUST bind authorization codes to a certain
   client and authenticate it using an appropriate mechanism (e.g.
   client credentials **OR** PKCE).

The statement below is a bit misleading from the previous as it states **client credentials OR PKCE** (not both).

Either way, I'm all for making things more secure. I'm very curious though which OAuth 2.0 Provider's have implemented this recommended best practice? Can you point me to a provider that supports PKCE for a confidential client (in a web app)? A link to the reference documentation would be helpful.

Our primary goal when building out new features is to be spec compliant first and foremost and to provide out-of-the-box support for extension features supported by **popular** OAuth 2.0 Providers. As well, to provide extension points for users to implement for edge-case scenarios.

Reply:
We have it implemented in our custom provider but that aside, did you see the example from oauth.com: https://oauth.com/playground/authorization-code-with-pkce.html
This exactly demonstrates the use of PKCE in a web application.

Reply:
@vanwobe @sdoxsee I have confirmed that Okta does indeed support PKCE for confidential clients. I have logged #6548 to track this feature enhancement. I'd rather get this PR merged based on the spec and than look at adding that support as a feature enhancement after.